### PR TITLE
Rewrite docker-checker.sh to make it less kill-happy.

### PR DESCRIPTION
**Pull request by mml**
Also a little more organized and easier to re-use.
Should minimize the incidence of #23461.

### CODE REVIEW

**Comment by mml** [comment]
Robot, y u no pick reviewer?

**Comment by mml** [comment]
ping @ArtfulCoder

**Comment by ArtfulCoder** [comment]
can you combine the init_healthy and health functions into one function that takes the timeout as a parameter.

**Comment by mml** [comment]
I think this looks a little harder to read, but LMK what you think.  I'll squash commits before merge.

File: cluster/saltbase/salt/supervisor/docker-checker.sh [hunk 1]
```diff
@@ -18,38 +18,70 @@
...
+# the number of seconds to keep trying for a healthy result.  If none is passed
+# we make only one attempt.
+healthy() {
+  max_retry_sec="$1"
+  shift
+
+  starttime=$(date +%s)
+  while ! timeout 60 docker ps > /dev/null; do
+    if [[ -z "$max_retry_sec" || $(( $(date +%s) - starttime )) -gt "$max_retry_sec" ]]; then
+      echo "Docker failed!"
```
Comment (ArtfulCoder) [inline cluster/saltbase/salt/supervisor/docker-checker.sh]: can you say docker ps command did not succeed.

File: cluster/saltbase/salt/supervisor/docker-checker.sh [hunk 2]
```diff
@@ -35,12 +35,17 @@ main() {
   exit 2
 }
 
-# Performs an "initial" health check on docker.  This loops until it gets a
-# healthy result for a max of 60s, at which point it will return nonzero.
-init_healthy() {
+# Performs health check on docker.  If a parameter is passed, it is treated as
+# the number of seconds to keep trying for a healthy result.  If none is passed
+# we make only one attempt.
+healthy() {
```
Comment (ArtfulCoder) [inline cluster/saltbase/salt/supervisor/docker-checker.sh]: we don't need the param to be optional, but its ok if you have it as optional.

**Comment by ArtfulCoder** [comment]
LGTM
I need to check one thing with @dchen1107 
Dawn: does the script require that docker be restarted atleast once ?

**Comment by mml** [comment]
Ping @dchen1107 pretty sure this is the biggest blocker of the submit queue right now.

**Comment by ArtfulCoder** [comment]
assigned to Dawn. 
@lavalamp You might want to see this through when you take over on-call tomorrow.

**Comment by dchen1107** [comment]
LGTM

**Pull request merged**